--- vinagre-3.22.0/plugins/rdp/vinagre-rdp-tab.c
+++ vinagre-3.22.0/plugins/rdp/vinagre-rdp-tab.c
@@ -930,7 +930,7 @@ frdp_authenticate (freerdp  *instance,
                                         TRUE,
                                         TRUE,
                                         TRUE,
-                                        20,
+                                        255,
                                         domain,
                                         username,
                                         password,
