--- vinagre-3.22.0/vinagre/vinagre-notebook.c
+++ vinagre-3.22.0/vinagre/vinagre-notebook.c
@@ -521,11 +521,6 @@ build_tab_label (VinagreNotebook *nb,
     GtkCssProvider *css;
 
     static const gchar button_style[] = "* {\n"
-        "-GtkButton-default-border : 0;\n"
-        "-GtkButton-default-outside-border : 0;\n"
-        "-GtkButton-inner-border: 0;\n"
-        "-GtkWidget-focus-line-width : 0;\n"
-        "-GtkWidget-focus-padding : 0;\n"
         "padding: 0;\n"
         "}";
 
